hack/build-and-push.sh

```diff
@@ -133,95 +180,325 @@ if [ "$SKIP_BUILD" == "" ]; then
+                found=true
+                break
+            else
+                echo "error: task bundle $task_bundle has migration artifact $artifact_ref, but the migration content is different: $filename" 1>&2
+                exit 1
+            fi
         fi
+    done <<<"$artifact_refs"
 
-        ANNOTATION_FLAGS=()
-        for annotation in "${ANNOTATIONS[@]}"; do
-            ANNOTATION_FLAGS+=("--annotate" "$(escape_tkn_bundle_arg "$annotation")")
-        done
+    if [ -n "$found" ]; then
+        return 0
+    fi
 
-        output=$(tkn_bundle_push "${ANNOTATION_FLAGS[@]}" -f "$prepared_task_file" "$task_bundle" | save_ref "$task_bundle" "$OUTPUT_TASK_BUNDLE_LIST")
-        echo "$output"
-        task_bundle_with_digest="${output##*$'\n'}"
+    (
+        cd "${migration_file%/*}"
+        retry oras attach \
+            --registry-config "$AUTH_JSON" \
+            --artifact-type "$ARTIFACT_TYPE_TEXT_XSHELLSCRIPT" \
+            --distribution-spec v1.1-referrers-tag \
+            --annotation "$ANNOTATION_TASK_MIGRATION=true" \
+            "$task_bundle" "${migration_file##*/}"
+    )
 
-        # copy task to new tag pointing to commit where the file was changed lastly, so that image persists
-        # even when original tag is updated
-        skopeo copy "docker://${task_bundle}" "docker://${task_bundle}-${task_file_sha}"
-    fi
-    # version placeholder is removed naturally by the substitution.
-    real_task_name=$(yq e '.metadata.name' "$prepared_task_file")
-    sub_expr_1="
-        (.spec.tasks[].taskRef | select(.name == \"${real_task_name}\" and .version == \"${task_version}\" ))
-        |= {\"resolver\": \"bundles\", \"params\": [ { \"name\": \"name\", \"value\": \"${real_task_name}\" } , { \"name\": \"bundle\", \"value\": \"${task_bundle_with_digest}\" }, { \"name\": \"kind\", \"value\": \"task\" }] }
-    "
-    sub_expr_2="
-        (.spec.finally[].taskRef | select(.name == \"${real_task_name}\" and .version == \"${task_version}\" ))
-        |= {\"resolver\": \"bundles\", \"params\": [ { \"name\": \"name\", \"value\": \"${real_task_name}\" } , { \"name\": \"bundle\", \"value\": \"${task_bundle_with_digest}\" },{ \"name\": \"kind\", \"value\": \"task\" }] }
-    "
-    for filename in "$generated_pipelines_dir"/*.yaml "$core_services_pipelines_dir"/*.yaml
+    echo
+    echo "Attached migration file ${migration_file} to ${task_bundle}"
+
+    return 0
+}
+
+build_push_tasks() {
+    local build_data
+    local task_bundle_with_digest
+    local migration_file
+    local prepared_task_file
+    local task_file_sha
+    local task_bundle
+
+    find task/*/* -maxdepth 0 -type d | awk -F '/' '{ print $0, $2, $3 }' | \
+    while read -r task_dir task_name task_version
     do
-        yq e "$sub_expr_1" -i "${filename}"
-        yq e "$sub_expr_2" -i "${filename}"
+        if [ -n "$TEST_TASKS" ] && echo "$TEST_TASKS" | grep -v "$task_name" 2>/dev/null; then
```

Comment:
```suggestion
        if [ -n "$TEST_TASKS" ] && echo "$TEST_TASKS" | grep -qv "$task_name" 2>/dev/null; then
```
When I set `TEST_TASKS=buildah`, the script was just printing `buildah` over and over again :smile: